src/inference/emcee.jl

```diff
@@ -0,0 +1,243 @@
+###
+### Sampler states
+###
+
+struct Emcee{space, E<:AMH.Ensemble} <: InferenceAlgorithm 
+    ensemble::E
+end
+
+function Emcee(n_walkers::Int, stretch_length=2.0)
+    # Note that the proposal distribution here is just a Normal(0,1)
+    # because we do not need AdvancedMH to know the proposal for
+    # ensemble sampling.
+    prop = AMH.StretchProposal(Normal(0, 1), stretch_length)
+    ensemble = AMH.Ensemble(n_walkers, prop)
+    return Emcee{(), typeof(ensemble)}(ensemble)
+end
+
+function Sampler(
+    alg::Emcee,
+    model::Model,
+    s::Selector=Selector()
+)
+    # Set up info dict.
+    info = Dict{Symbol, Any}()
+
+    # Set up state struct.
+    state = SamplerState(VarInfo(model))
+
+    # Generate a sampler.
+    return Sampler(alg, info, s, state)
+end
+
+alg_str(::Sampler{<:Emcee}) = "Emcee"
+
+function AbstractMCMC.sample_init!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:Emcee},
+    N::Integer;
+    verbose::Bool=true,
+    resume_from=nothing,
+    kwargs...
+)
+    # Resume the sampler.
+    set_resume!(spl; resume_from=resume_from, kwargs...)
+
+    # Get `init_theta`
+    initialize_parameters!(spl; verbose=verbose, kwargs...)
+
+    # Link everything before sampling.
+    link!(spl.state.vi, spl)
+end
+
+function AbstractMCMC.sample_end!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:Emcee},
+    N::Integer,
+    transitions;
+    kwargs...
+)
+    # Invlink everything after sampling.
+    invlink!(spl.state.vi, spl)
+end
+
+function AbstractMCMC.step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:Emcee},
+    N::Integer,
+    transition::Nothing;
+    kwargs...
+)
+    # Generate a log joint function.
+    densitymodel = AMH.DensityModel(gen_logπ(spl.state.vi, spl, model))
+
+    # Make the first transition.
+    transition = sample(rng, model, Prior(), spl.alg.ensemble.n_walkers, chain_type=Any, progress=false)
+
+    ss = DynamicPPL.SampleFromPrior()
+    walkers = Vector{AMH.Transition}(undef, spl.alg.ensemble.n_walkers)
+
+    for (i, t) in enumerate(transition)
+        DynamicPPL.invlink!(spl.state.vi, spl)
+        spl.state.vi[spl] = identity.(t[ss])
+        DynamicPPL.link!(spl.state.vi, spl)
+        walkers[i] = AMH.Transition(spl.state.vi[spl], getlogp(spl.state.vi))
+    end
```

Comment:
Is it possible to not preallocate `walkers`? It results in a vector with abstract element type which we could avoid by using, e.g., `map`, I guess. Also I guess it would be nice if we could get rid of the use of `sample`, `invlink!`, and `link!` here. Can't we just run the model multiple times (using the default `SampleFromPrior`) and create `AMH.Transition`s with the resulting variables?

Reply:
Yeah, I can do that.